### PR TITLE
release/snowplow-web/0.1.0

### PR DESCRIPTION
**Pull request by bill-warner**
Release of snowplow-web v0.1.0-beta.0.

- Only supports Redshift.
- Uses incremental hooks provided by snowplow-utils package
- Includes integration tests with an anonymised dummy data
  - Hashed fields: `domain_userid`, `network_userid`, `domain_sessionid`, `user_id`, `user_ipaddress`
- Includes `custom_example` directory, which contains a dummy dbt project with custom modules built out.
- Doc site hosted on Git pages.

### CODE REVIEW

**Comment by paulboocock** [comment]
Quick note: Before we merge and release/announce this anywhere, I'd like to move it to `snowplow`. We know this is going to be an important part of our offering going forward so I think we should put it in the `snowplow` org from the get go, even during the beta phase (for simplicities sake). Same for the `utils` package too.

**Comment by bill-warner** [comment]
> Quick note: Before we merge and release/announce this anywhere, I'd like to move it to `snowplow`. We know this is going to be an important part of our offering going forward so I think we should put it in the `snowplow` org from the get go, even during the beta phase (for simplicities sake). Same for the `utils` package too.

Yeh I agree. When moving, I think we should rename the repo to `dbt-snowplow-web` to align with the [naming convention](https://docs.getdbt.com/docs/guides/building-packages#2-create-your-new-project) for repos containing dbt packages. I missed this point when creating this repo :(

File: models/users/manifest/snowplow_web_users_manifest.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    disable_upsert_lookback=true,
+    full_refresh=false,
+    schema=var("snowplow__manifest_custom_schema"),
+    sort='start_tstamp',
+    dist='domain_userid'
+  ) 
+}}
+
+-- TODO: Consider retiring this model by adapting sessions_lifecycle_manifest to include `domain_userid`.
+-- Think through edge case of multiple `domain_userid` per session
```
Comment (colmsnowplow) [inline models/users/manifest/snowplow_web_users_manifest.sql:15]: Is there a legitimate case where a session can have two domain_userids?

File: docs/run_results.json [hunk 2]
```diff
@@ -0,0 +1 @@
...tart_tstamp"}, {"status": "success", "timing": [{"name": "compile", "started_at": "2021-07-07T13:18:48.198111Z", "completed_at": "2021-07-07T13:18:48.204085Z"}, {"name": "execute", "started_at": "2021-07-07T13:18:48.204241Z", "completed_at": "2021-07-07T13:18:48.204248Z"}], "thread_id": "Thread-1", "execution_time": 0.0064640045166015625, "message": null, "adapter_response": {}, "unique_id": "test.snowplow_web.unique_snowplow_web_users_domain_userid"}, {"status": "success", "timing": [{"name": "compile", "started_at": "2021-07-07T13:18:48.204762Z", "completed_at": "2021-07-07T13:18:49.059867Z"}, {"name": "execute", "started_at": "2021-07-07T13:18:49.060400Z", "completed_at": "2021-07-07T13:18:49.060419Z"}], "thread_id": "Thread-1", "execution_time": 0.925055980682373, "message": null, "adapter_response": {}, "unique_id": "test.snowplow_web_integration_tests.dbt_utils_equality_snowplow_web_users_actual_ref_snowplow_web_users_expected_it_"}], "elapsed_time": 13.01059103012085, "args": {"log_format": "default", "write_json": true, "profiles_dir": "/Users/williamwarner/.dbt", "use_cache": true, "compile": true, "version_check": true, "which": "generate", "rpc_method": "docs.generate"}}
```
Comment (colmsnowplow) [inline docs/run_results.json:1]: What is this file? Should it be included in a package's source code?

File: integration_tests/.scripts/integration_test_w_custom_module.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+echo "Snowplow web integration tests: Run 1: standard modules"
+
+eval "dbt run --target $DATABASE --full-refresh --vars 'teardown_all: true'" || exit 1;
+
+echo "Snowplow web integration tests: Run 2: standard modules"
+
+eval "dbt run --target $DATABASE" || exit 1;
+
+echo "Snowplow web integration tests: Run 3: Partial backfill of custom module + standard modules"
+# This tests the functionality of the snowplow_utils.is_run_with_new_events() macro
```
Comment (colmsnowplow) [inline integration_tests/.scripts/integration_test_w_custom_module.sh:26]: Ideally, we assume that this macro does its job correctly, but have a thorough test for it in the snowplow-dbt-utils repo.

File: README.md [hunk 4]
```diff
@@ -1 +1,69 @@
...
+### Requirements
+- A dataset of web events from the [Snowplow JavaScript tracker][tracker-docs] must be available in the database.
+- Have the [`webPage` context][webpage-context] enabled.
+
+### Installation
+Check dbt Hub for the latest installation instructions, or read the [dbt docs][dbt-package-docs] for more information on installing packages.
+
+### Configuration & Operation
+Please refer to the [doc site][snowplow-web-docs] for details on how to configure and run the package.
+
```
Comment (colmsnowplow) [inline README.md:26]: Consider adding a brief note on testing perhaps?

**Review by colmsnowplow — APPROVED** [review]
Looks good to me! Great stuff.

I think we can clean up some unnecessary files - for example `.gitkeep` files need only exist when the rest of the directory is empty.

File: docs/run_results.json (hunk 2 above)
Comment (bill-warner) [reply docs/run_results.json:1]: I believe it is required to build the doc site.

https://docs.getdbt.com/docs/guides/building-packages#5-deploy-the-docs-for-your-package

File: integration_tests/.scripts/integration_test_w_custom_module.sh (hunk 3 above)
Comment (bill-warner) [reply integration_tests/.scripts/integration_test_w_custom_module.sh:26]: Fair point. I am building out a series of tests for the macros in the utils package atm so should be able to cover this off in the next release.

File: models/users/manifest/snowplow_web_users_manifest.sql (hunk 1 above)
Comment (bill-warner) [reply models/users/manifest/snowplow_web_users_manifest.sql:15]: Well [this ticket](https://snowplow.zendesk.com/agent/tickets/27235) reported such behaviour a while back although I am not sure if we ever got to the root cause. At the time, I checked our RS prod db to see if I could find the same case and came across such an instance:

```sql
WITH page_view_context AS (
SELECT
    *
FROM atomic.com_snowplowanalytics_snowplow_web_page_1
WHERE DATE(root_tstamp) = '2021-06-05'
)
SELECT
    a.app_id,
    a.collector_tstamp,
    a.derived_tstamp,
    a.event_name,
    a.domain_sessionid,
    a.domain_userid,
    b.id AS page_view_id
FROM atomic.events a
LEFT JOIN page_view_context b
ON a.collector_tstamp = b.root_tstamp
AND a.event_id = b.root_id
WHERE DATE(a.collector_tstamp) = '2021-06-05'
AND a.domain_sessionid = '85366e54-3b76-4577-ae36-cebd42f659f5'
ORDER BY 2
```
It certainly isnt a legitimate case however. I think I will explore removing this model in the BQ release. I am keen to minimise the amount of incremental manifest tables.

File: README.md (hunk 4 above)
Comment (bill-warner) [reply README.md:26]: Which tests exactly? Running tests on the models within the package?

Comment (colmsnowplow) [reply README.md:26]: I was thinking within the package. If the readme is intended for people who wouldn't typically run those tests though it doesn't need it.

File: models/users/manifest/snowplow_web_users_manifest.sql (hunk 1 above)
Comment (colmsnowplow) [reply models/users/manifest/snowplow_web_users_manifest.sql:15]: Yeah agreed on both counts

**Pull request merged**